### PR TITLE
Puddles & spreader refactor

### PR DESCRIPTION
**Pull request by metalgearsloth**
Requires https://github.com/space-wizards/RobustToolbox/pull/3912

I have little idea how smoke functions ingame so there's a high likelihood it may break for reactions with entities so someone more familiar with it should check.

Mopping changes:
- Mopping is based around water transfer. Using a mop on a puddle transfers water into it and takes non-water out. Water puddles will evaporate over time and leave sparkles. This also means if a janitor is not careful they might combine water with something unintended for fun surprises.
- Sparkles spawn when the puddle fully evaporates. An examine will also tell you if it's evaporating or not.
- Mopping uses interaction cooldown instead of do_afters as it doesn't need to be uninterruptible and feels better (though no prediction). Advanced mop can hold more and has a shorter cooldown.
- Mop uses the same splitbar as gas analzyer to show reagents in itemstatus so you know water vs non-water.

Spreader changes:
- Smoke, kudzu, and puddles all use the same spreader logic.
- Smoke made me want to die. I refactored it entirely to be vaguely similar. I don't even understand if the touch interactions work the same (I think uhh foam did touch contact but I made all smoke just do ingestion?)
- Spreading is now per group. ATM it's just a flat rate per group though any systems subscribed can implement their own custom update rate.
- Puddles overflow is a bit more randomised. It also doesn't use raycast as using it with nodes was too expensive for the current rebuild logic (300ms+ for ventclog event).
- It's still slightly annoying to use as you have to bookkeep EdgeSpreaderComponent slightly (i.e. adding / removing it as you require) and handle your own spread logic (i.e. setting up solutions / kudzu growth yourself) but it's significantly better than the 3 systems doing their own thing and the event gives you your neighbours, plus the systems have access to how many nodes are in a group easily.

Puddles also now use iconsmoothing and use different sprite states instead of alpha to convey volume.
This likely still needs work but most of the hard stuff (cough smoke) is done.

Mop buckets can transfer reagent from drainable stuff (e.g. water / fuel tanks) to themselves via drag-drop. They can also transfer to a drain via drag-drop to empty them.

https://user-images.githubusercontent.com/31366439/230556727-d5b85193-12af-4e65-b266-b238d3f99866.mp4

:cl: metalgearsloth & Aleksh
- tweak: Mops now use a cooldown instead of a do_after.
- tweak: Mops now transfer water onto tiles where it will then evaporate over time.
- tweak: Mop ItemStatus UI now shows the water vs non-water vs empty reagents so you know how much space you have.
- add: New smoothed puddle sprites by Aleksh.
- tweak: Puddle sparkles spawn when the puddle finishes evaporating.
- add: Buckets can now be dragged onto tanks to refill them and onto drains to empty them.

### CODE REVIEW

**Comment by Myakot** [comment]
Praise be this PR, this is so QoL.

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 1]
```diff
@@ -1,140 +1,391 @@
...
+            NeighborOccupiedTiles = occupiedTiles,
+            Neighbors = neighbors,
+            Updates = updates,
+        };
+
+        RaiseLocalEvent(uid, ref ev);
+        updates = ev.Updates;
+    }
+
+    public void GetNeighbors(EntityUid uid, string groupName, out ValueList<Vector2i> freeTiles, out ValueList<Vector2i> occupiedTiles, out ValueList<EntityUid> neighbors)
```
Comment (moonheart08) [inline Content.Server/Kudzu/SpreaderSystem.cs:231]: where doc

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 2]
```diff
@@ -1,140 +1,391 @@
...
                 break;
+            }
+
+            if (oldCount == occupiedTiles.Count)
+                freeTiles.Add(neighborPos);
         }
     }
 
-    private bool TryGrow(EntityUid ent, TransformComponent? transform = null, SpreaderComponent? spreader = null)
+    public List<EntityUid> GetNeighbors(EntityUid uid, AirtightComponent comp)
```
Comment (moonheart08) [inline Content.Server/Kudzu/SpreaderSystem.cs:314]: wher doc

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 3]
```diff
@@ -1,140 +1,391 @@
...
-            return true;
+/// <summary>
+/// Raised every tick to determine how many updates a particular spreading node group is allowed.
+/// </summary>
+[ByRefEvent]
+public record struct SpreadGroupUpdateRate()
+{
+    public string Name;
+    public int UpdatesPerSecond = 16;
+}
```
Comment (moonheart08) [inline Content.Server/Kudzu/SpreaderSystem.cs:364]:
```suggestion
public record struct SpreadGroupUpdateRate(string Name, int UpdatesPerSecond = 16);
```

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 4]
```diff
@@ -1,140 +1,391 @@
...
 
-        if (!EntityManager.TryGetComponent<AirtightComponent>(ent, out var airtight))
-            return false;
+[RegisterComponent]
+public sealed class SpreaderGridComponent : Component
+{
+    [DataField("nextUpdate", customTypeSerializer:typeof(TimeOffsetSerializer))]
+    public TimeSpan NextUpdate = TimeSpan.Zero;
+}
```
Comment (moonheart08) [inline Content.Server/Kudzu/SpreaderSystem.cs:371]: own file bls

File: Content.Shared/Chemistry/Components/Solution.cs [hunk 5]
```diff
@@ -432,6 +432,15 @@ public void RemoveAllSolution()
             _heatCapacity = 0;
         }
 
+        public Solution SplitSolutionWithout(FixedPoint2 toTake, string without)
```
Comment (moonheart08) [inline Content.Shared/Chemistry/Components/Solution.cs:435]: where mydoc

File: Content.Shared/Fluids/Components/DrainComponent.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.Fluids.Components;
+
+[RegisterComponent, NetworkedComponent]
+public sealed class DrainComponent : Component
```
Comment (moonheart08) [inline Content.Shared/Fluids/Components/DrainComponent.cs:6]: doc comment

File: Content.Shared/Fluids/Components/DrainComponent.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public const string SolutionName = "drainBuffer";
+
+    [DataField("accumulator")]
+    public float Accumulator = 0f;
```
Comment (moonheart08) [inline Content.Shared/Fluids/Components/DrainComponent.cs:11]: what's this accumulating?? (doc comment)

File: Content.Shared/Fluids/SharedAbsorbentSystem.cs [hunk 8]
```diff
@@ -1,9 +1,10 @@
+using System.Linq;
 using Robust.Shared.GameStates;
 using Robust.Shared.Serialization;
 
 namespace Content.Shared.Fluids;
 
-public abstract class SharedMoppingSystem : EntitySystem
+public abstract class SharedAbsorbentSystem : EntitySystem
```
Comment (moonheart08) [inline Content.Shared/Fluids/SharedAbsorbentSystem.cs:7]: ock comment

File: Content.Shared/Fluids/SharedPuddleSystem.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
+using Content.Shared.Chemistry.Components;
+using Content.Shared.DragDrop;
+using Content.Shared.Fluids.Components;
+
+namespace Content.Shared.Fluids;
+
+public abstract class SharedPuddleSystem : EntitySystem
```
Comment (moonheart08) [inline Content.Shared/Fluids/SharedPuddleSystem.cs:7]: dock comments

File: Content.Shared/Fluids/SharedPuddleSystem.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+{
+    public const float LowThreshold = 0.3f;
+
+    public override void Initialize()
```
Comment (moonheart08) [inline Content.Shared/Fluids/SharedPuddleSystem.cs:11]:
```suggestion
    /// <inheritdoc/>
    public override void Initialize()
```

File: Content.Shared/Fluids/SharedPuddleSystem.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
+using Content.Shared.Chemistry.Components;
+using Content.Shared.DragDrop;
+using Content.Shared.Fluids.Components;
+
+namespace Content.Shared.Fluids;
+
+public abstract class SharedPuddleSystem : EntitySystem
+{
+    public const float LowThreshold = 0.3f;
```
Comment (moonheart08) [inline Content.Shared/Fluids/SharedPuddleSystem.cs:9]: this especially needs a :sparkles:doc comment:sparkles:

**Review by moonheart08 — CHANGES_REQUESTED** [review]
ftjbmkmjrdmknbcm 

a crumb of documentation

File: Content.Server/Kudzu/KudzuSystem.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            args.Updates--;
+
+            if (args.Updates <= 0)
+                return;
+        }
+    }
+
+    private void OnKudzuUpdateRate(ref SpreadGroupUpdateRate args)
+    {
+        if (args.Name != "kudzu")
```
Comment (mirrorcult) [inline Content.Server/Kudzu/KudzuSystem.cs:58]: if this is in c# only can these just be an enum

File: Content.Shared/Fluids/Components/DrainComponent.cs (hunk 7 above)
Comment (mirrorcult) [reply Content.Shared/Fluids/Components/DrainComponent.cs:11]: pretty sure this is just a copy from old code

**Review by mirrorcult — APPROVED** [review]
github was beating my ass trying to review this PR

works fine in game to the best of my testing ability

File: Content.Server/Fluids/EntitySystems/PuddleSystem.Evaporation.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+public sealed partial class PuddleSystem
+{
+    private static readonly TimeSpan EvaporationCooldown = TimeSpan.FromSeconds(1);
+
+    /// <summary>
+    /// Assuming puddle is 20 units then ~1 minute for it to evaporate.
+    /// </summary>
+    private static readonly FixedPoint2 EvaporationAmount = FixedPoint2.New(0.3);
+
+    public const string EvaporationReagent = "Water";
```
[the remark this answers is not in the document]
Comment (mirrorcult) [reply Content.Server/Fluids/EntitySystems/PuddleSystem.Evaporation.cs:17]: this is fine for now i'm working on a system for this that will replace it

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 14]
```diff
@@ -1,140 +1,391 @@
...
-    private float _accumulatedFrameTime = 0.0f;
+    private static readonly TimeSpan SpreadCooldown = TimeSpan.FromSeconds(1);
 
-    private readonly HashSet<EntityUid> _edgeGrowths = new ();
+    private static readonly List<string> SpreaderGroups = new()
+    {
+        "kudzu",
+        "puddle",
+        "smoke",
+    };
```
[the remark this answers is not in the document]
Comment (mirrorcult) [reply Content.Server/Kudzu/SpreaderSystem.cs:89]: or an enum at the very least

File: Content.Server/Fluids/EntitySystems/PuddleSystem.cs [hunk 15]
```diff
@@ -1,313 +1,621 @@
...
 using Robust.Shared.Player;
 using Solution = Content.Shared.Chemistry.Components.Solution;
 using Robust.Shared.Prototypes;
 using Robust.Shared.Random;
+using Robust.Shared.Timing;
 
-namespace Content.Server.Fluids.EntitySystems
+namespace Content.Server.Fluids.EntitySystems;
+
+public sealed partial class PuddleSystem : SharedPuddleSystem
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Fluids/EntitySystems/PuddleSystem.cs:30]: Already on the other partial.

File: Content.Server/Fluids/EntitySystems/PuddleSystem.cs [hunk 16]
```diff
@@ -1,313 +1,621 @@
...
+    [Dependency] private readonly AudioSystem _audio = default!;
+    [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly ReactiveSystem _reactive = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+    [Dependency] private readonly SharedPopupSystem _popups = default!;
+    [Dependency] private readonly StepTriggerSystem _stepTrigger = default!;
+    [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+
+    public static float PuddleVolume = 1000;
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Fluids/EntitySystems/PuddleSystem.cs:46]: mirror issue.

File: Content.Server/Fluids/EntitySystems/SmokeSystem.cs [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+        if (component.SpreadAmount == 0 ||
+            args.Grid == null ||
+            !_solutionSystem.TryGetSolution(uid, SmokeComponent.SolutionName, out var solution) ||
+            args.NeighborFreeTiles.Count == 0)
+        {
+            RemCompDeferred<EdgeSpreaderComponent>(uid);
+            return;
+        }
+
+        var prototype = MetaData(uid).EntityPrototype;
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Fluids/EntitySystems/SmokeSystem.cs:84]: This is how it was in the old system, yes, and I just wanted puddles.

File: Content.Server/Kudzu/KudzuSystem.cs [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+            return;
+        }
+
+        if (args.NeighborFreeTiles.Count == 0 || args.Grid == null)
+        {
+            RemCompDeferred<EdgeSpreaderComponent>(uid);
+            return;
+        }
+
+        var prototype = MetaData(uid).EntityPrototype?.ID;
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Kudzu/KudzuSystem.cs:34]: me too.

File: Content.Server/Kudzu/SpreaderSystem.cs [hunk 19]
```diff
@@ -1,140 +1,391 @@
...
     [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
 
-    /// <summary>
-    /// Maximum number of edges that can grow out every interval.
-    /// </summary>
-    private const int GrowthsPerInterval = 1;
-
-    private float _accumulatedFrameTime = 0.0f;
+    private static readonly TimeSpan SpreadCooldown = TimeSpan.FromSeconds(1);
```
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Kudzu/SpreaderSystem.cs:82]: This one's a lot of effort for something mirror is about to refactor.

File: Content.Shared/Fluids/Components/DrainComponent.cs (hunk 7 above)
Comment (metalgearsloth) [reply Content.Shared/Fluids/Components/DrainComponent.cs:11]: it is yeah I just moved it to shared.

Comment (metalgearsloth) [reply Content.Shared/Fluids/Components/DrainComponent.cs:11]: This is also why none of it's networked (I just need it in shared for dragdrop to work).

**Pull request merged**